Refactor IPv6 DAD timer

jukkar opened the pull request:
This is the similar thing that was done for lifetime timer. This saves > 20 bytes / allocated IPv6 address for each network interface. I also added tests for testing DAD timers as that was missing.

jukkar committed:
net: if: Refactor IPv6 address lifetime timer

In preparation of DAD timer rework, refactor IPv6 address lifetime
timer setting in net_if_addr.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: if: Convert to use only one IPv6 DAD timer

Similar change as what was done for address lifetime work
earlier. This will save over 20 bytes of memory for each
unicast IPv6 address for each network interface in the system.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
tests: net: ipv6: Add tests for verifying DAD timers

Make sure that DAD timers are triggered in proper order.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

pfalcon commented on include/net/net_if.h:
```diff
@@ -64,7 +64,16 @@ struct net_if_addr {
 	/** Duplicate address detection (DAD) timer */
-	struct k_delayed_work dad_timer;
+	struct {
+		/** Used for track timers */
```
"to track"

pfalcon commented on subsys/net/ip/net_if.c:
```diff
@@ -555,6 +561,90 @@ static void dad_timeout(struct k_work *work)
+static bool dad_check_timeout(s64_t start, u32_t time, s64_t timeout)
+{
+	start += time;
+	start = abs(start);
```
Using signed 64bit numbers for always-increasing time is indeed confusing...

jukkar replied:
The k_uptime_get() returns s64_t, I am not sure what is the point in your comment here.

pfalcon replied:
k_uptime_get() returns s64_t, but is described as "This routine returns the elapsed time since the system booted". It's clear that elapsed time goes only forward, and cannot be negative. At one point in the unattainable future it will overflow, but C standard says that the result of signed overflow is undefined, so a good C programmer cannot do anything about it anyway. But maybe you personally don't care about the C standard, and assume that signed numbers are 2's-complement. But using abs() in this case doesn't make any sense, because at the point of overflow, such abs'ed signed time will go backwards. Let me demonstrate this on 16-bit numbers:

After 32768, there will be -32768. Applying abs() to it will lead to the same number. But then there will be -32767, which abs will turn into 32767, etc. I.e. the "time" on which you perform calculations will go as follows:

32766, 32767, -32768, 32767, 32766, 32765, ... 10, 9, 8, ...

So, using, abs() totally doesn't make sense. And that's after performing signed operations on wrap-around time, which don't make sense first.

jukkar replied:
Your comments are in weird order in github and this gets quite convoluted.

Anyway, about the conversions. We are only interested in difference of the numbers and at least gcc returns the difference correctly as shown by this example

```
main()
{
	int32_t max_int = INT32_MAX;
	int32_t max_int2 = INT32_MAX + 2;
	int32_t diff = max_int - max_int2;

	printf("max sint %d\n", max_int);
	printf("max + 2 sint %d\n", max_int2);
	printf("diff %d\n", diff);
}
```
will print
```
max sint 2147483647
max + 2 sint -2147483647
diff -2
```

Taking abs() from the diff makes sense here.

pfalcon commented on subsys/net/ip/net_if.c:
```diff
@@ -555,6 +561,90 @@ static void dad_timeout(struct k_work *work)
+		return UINT32_MAX;
+	}
+
+	next_timeout = timeout - (ifaddr->dad.timer_start +
```
Very confusing arithmetics, mixing up 32 and 64 bit numbers, signed and unsigned.

jukkar replied:
We just need the diff of two s64_t values, and that diff will fit s32_t

pfalcon replied:
No, there're no s64_t values! There're only u64_t values disguised as s64_t, as explained in the previous comment. Please avoid using hosed k_uptime_get(), instead use sanely-defined k_uptime_get_32() (wraparound period 49.7 days, more than enough for networking needs), and leave all these problems behind.

pfalcon commented on subsys/net/ip/net_if.c:
```diff
@@ -555,6 +561,90 @@ static void dad_timeout(struct k_work *work)
+				 timeout);
+}
+
+static u32_t dad_manage_timeouts(struct net_if_addr *ifaddr, s64_t timeout)
```
Why this is called manage_timeout**s**, while it's apparently deals with a single one?

pfalcon replied:
Then in should be named like "manage_and_return_timeouts". But as long as it manages a *single* timeout, the naming is still confusing ;-).

pfalcon commented on subsys/net/ip/net_if.c:
```diff
@@ -555,6 +561,90 @@ static void dad_timeout(struct k_work *work)
+		}
+	}
+
+	if (timeout_update < (UINT32_MAX - 1)) {
```
Apparently, here should be check for UINT32_MAX, it's the value which has a special meaning in dad_manage_timeouts() retval.

pfalcon replied:
Can you please elaborate how the entire processing works then?

jukkar replied:
The timer_manage_timeout() will return UINT32_MAX if there was a timeout. The initial value is set to UINT32_MAX-1 so that we will know if there are any timers that should be triggered. The timeout_update will eventually contain the lowest timeout value that is found in the list. If there was none, then timeout_update will contain either UINT32_MAX or UINT32_MAX-1 value and no work needs to be triggered.

pfalcon replied:
Thanks for the detailed description, just what I was looking for! But my [rubber ducky](https://en.wikipedia.org/wiki/Rubber_duck_debugging) sees it differently:

Indeed, timer_manage_timeout() returns either UINT32_MAX to signify that timer already expired, or any other value (i.e. in the range 0..UINT32_MAX-1) to signify delay until the timer expiration. I.e. conceptually, it returns Union(0..UINT32_MAX-1, MAGIC_VALUE=UINT32_MAX).

Now it's caller, dad_timeout(), should also either deal with magic value of UINT32_MAX, or with a value in range 0..UINT32_MAX-1. But instead, it introduces another magic value, UINT32_MAX-1. Interpretation of that value appears to be the same as for UINT32_MAX, but there's a bug - if timer_manage_timeout() returns UINT32_MAX-1, which from its PoV is a normal timer delay, then dad_timeout() will interpret it as a magic value. But the main problem is that there's no need for another magic value at all, so only UINT32_MAX should be used throughout.

jukkar replied:
Yes, I noticed it too, the -1 magic is gone already in the latest version.

pfalcon commented on subsys/net/ip/net_if.c:
```diff
@@ -555,6 +561,90 @@ static void dad_timeout(struct k_work *work)
+
+static void dad_timeout(struct k_work *work)
+{
+	u32_t timeout_update = UINT32_MAX - 1;
```
Can (and should, to avoid confusion) be UINT32_MAX, similar or comment below.

jukkar replied:
Yes, that can be done

jukkar committed:
net: if: Refactor common timer functions

The DAD and adderess lifetime timers have common functionality
so refactor them to have only one set of timer functions. This
reduces used memory for code.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

pfalcon commented on subsys/net/ip/net_if.c:
```diff
@@ -555,6 +561,90 @@ static void dad_timeout(struct k_work *work)
+	if (timeout_update < (UINT32_MAX - 1)) {
+		NET_DBG("Waiting for %u ms", timeout_update);
+
+		k_delayed_work_submit(&address_lifetime_timer, timeout_update);
```
It's a bit strange that a routine related to DAD, with dad_timer available, submits work against address_lifetime_timer. Please add a code comment explaining why.

pfalcon replied:
#2 issue: you pass unsigned value to the function which expects signed one (k_delayed_work_submit). The right solution is of course to fix k_delayed_work_submit(), which takes s32_t, but deals with it as if it was u32_t (violating C standard in regard to overflow handling, etc.).

pfalcon reviewed (changes requested):
With 4-5 refactors like this already having gone in, and me being unable to follow thru them fully, I finally put anything else aside to delve into it. Unfortunately, I still can't see clearly what happens there, as the code is rather confusing, and needs at least commenting, if not cleanup.

jukkar commented:
Fixed the sanity check error and added more refactoring in the "net: if: Refactor common timer functions" commit.

jukkar commented:
Updated according to comments.

jukkar commented:
New version uploaded. Used wrong type (u64_t) for address lifetime timeout value, this cannot be done as the zephyr work api only uses 32 bit value.

pfalcon commented:
So, I made a detailed case for using u32_t and k_uptime_get_32(). We'll have to refactor to use them eventually, 64-bit values are superfluous. Can you give a direct answer why you don't want to do that now?

pfalcon commented on subsys/net/ip/net_if.c:
```diff
@@ -527,16 +535,49 @@ static void leave_mcast_all(struct net_if *iface)
+	}
+
+	next_timeout = current_time - (timer_start + timer_timeout);
+	return abs(next_timeout);
```
As argued yesterday, no need to use abs() when working with events ordered in time in known way. And as there's no need for them, then if you have them, that's highly confusing. Sorry, I'm firm on this, and that's the reason for -1.

Here, you tested timer is not timed out. So, you know that current_time is less than (timer_start + timer_timeout). So, just use unsigned arithmetics and expression of `(timer_start + timer_timeout) - current_time`.

jukkar commented on include/net/net_if.h:
```diff
@@ -50,19 +50,33 @@ struct net_if_addr {
+		/** DAD timer timeout. The timeout value is very short, so use
+		 * 16-bit value here.
+		 */
+		u16_t timer_timeout;
```
> Can you give a direct answer why you don't want to do that now?

I did not say I do not want to do it, I am investigating this atm. Instead of these kind of non-constructive comments, I would suggest that you start to send patches too if you find issues.

pfalcon replied:
Please assume that I'm trying to be constructive, in particular my aim is to get rid of my -1 on this and another review. So, I'd like to know if I should just wait while you with @rveerama1 consider u32_t option (a good, strategic decision), or if instead I should concentrate on tactical issues with the patches as it is (and then my concern remains use of abs()).

All in all, some more feedback exchange would help - I don't want to be looking as I'm blocking these refactors by throwing too far-fetched ideas (IMHO they aren't, but I can't know how you feel about it, unless you tell me). Sorry for confusion and hasting, if it looks like that.

pfalcon replied:
Re: sending patches, I'm indeed in the sad situation when I spend too much time on reviews and barely can work on development myself. Well, I always told that we don't have enough knowledge sharing among IP stack developers, and asked to be a default reviewers just to address that, so I keep telling myself that I should eat that dogfood for yet another week ;-). (Until my boss asks me "what the heck are you working on, where's your code"? ;-) ).

jukkar commented:
Moving the dad timer test patch into a separate pr #9549 as it is not really related to this one.

jukkar commented:
Most of this functionality will be replaced by #9659 so closing this one.
